### PR TITLE
WIP: Upgrade to 2.2.0

### PR DESCRIPTION
**Pull request by caleblloyd**
- Upgrade `Microsoft.EntityFrameworkCore.Relational` to 2.2.0
- Leave `Microsoft.EntityFrameworkCore.Relational.Specification.Tests` test suite at 2.1.4; taking that to 2.2.0 is a heavy-lift upgrade

### CODE REVIEW

**Comment by caleblloyd** [comment]
Project is building now, but test failures need to be investigated:

Functional Tests:
```
netcoreapp2.2: Total tests: 7464. Passed: 7221. Failed: 51. Skipped: 192.
netcoreapp2.0: Total tests: 7426. Passed: 7212. Failed: 22. Skipped: 192.
```

Integration Tests
```
netcoreapp2.2: Total tests: 38. Passed: 10. Failed: 28. Skipped: 0.
```

**Comment by FlyingWraptor** [comment]
Any ETA on a possible merge?
I'm happy to help to get this done ASAP.

**Comment by caleblloyd** [comment]
Failing tests need to be investigated and fixed.  You can open PRs with fixes to the `f_upgrade_2.2.0` branch.

**Comment by FlyingWraptor** [comment]
Will check it out in a minute

File: test/EFCore.MySql.Tests/Migrations/ModelSnapshotMySqlTest.cs [hunk 1]
```diff
@@ -159,11 +159,24 @@ protected void Test(Action<ModelBuilder> buildModel, string expectedCode, Action
...
+                new CSharpMigrationsGeneratorDependencies(
+                    codeHelper,
+                    new CSharpMigrationOperationGenerator(
+                        new CSharpMigrationOperationGeneratorDependencies(
+                            codeHelper)),
+                    new CSharpSnapshotGenerator(
+                        new CSharpSnapshotGeneratorDependencies(
+                            codeHelper))));
+
+            var code = generator.GenerateSnapshot("RootNamespace", typeof(DbContext), "Snapshot", model);
```
Comment (FlyingWraptor) [inline test/EFCore.MySql.Tests/Migrations/ModelSnapshotMySqlTest.cs]: It seems like you've started using the CSharpMigrationsGenerator but did not take the tests into consideration, it is now comparing just the builder code to an 'entire file of code'.

Testing:
```csharp
// <auto-generated />
 using System;
 using Microsoft.EntityFrameworkCore;
 using Microsoft.EntityFrameworkCore.Infrastructure;
 using Microsoft.EntityFrameworkCore.Storage.ValueConversion;

 namespace RootNamespace
 {
     [DbContext(typeof(DbContext))]
     partial class Snapshot : ModelSnapshot
     {
         protected override void BuildModel(ModelBuilder modelBuilder)
         {
 #pragma warning disable 612, 618
             modelBuilder
                 .HasAnnotation("ProductVersion", "2.2.0-rtm-35687")
                 .HasAnnotation("Relational:MaxIdentifierLength", 64);

             modelBuilder.Entity("Pomelo.EntityFrameworkCore.MySql.Migrations.ModelSnapshotMySqlTest+DataTypesVariable", b =>
                 {
                     b.Property<int>("Id")
                         .ValueGeneratedOnAdd();

                     b.Property<byte[]>("TypeByteArray")
                         .IsRequired();

                     b.Property<byte[]>("TypeByteArray255")
                         .IsRequired()
                         .HasMaxLength(255);

                     b.Property<byte[]>("TypeByteArray255N")
                         .HasMaxLength(255);

                     b.Property<byte[]>("TypeByteArrayN");

                     b.Property<string>("TypeJsonArray")
                         .IsRequired()
                         .HasColumnType("json");

                     b.Property<string>("TypeJsonArrayN")
                         .HasColumnType("json");

                     b.Property<string>("TypeJsonObject")
                         .IsRequired()
                         .HasColumnType("json");

                     b.Property<string>("TypeJsonObjectN")
                         .HasColumnType("json");

                     b.Property<string>("TypeString")
                         .IsRequired();

                     b.Property<string>("TypeString255")
                         .IsRequired()
                         .HasMaxLength(255);

                     b.Property<string>("TypeString255N")
                         .HasMaxLength(255);

                     b.Property<string>("TypeStringN");

                     b.HasKey("Id");

                     b.ToTable("DataTypesVariable");
                 });
 #pragma warning restore 612, 618
         }
     }
 }
```

vs the old
```csharp
builder
     .HasAnnotation("Relational:MaxIdentifierLength", 64);

 builder.Entity("Pomelo.EntityFrameworkCore.MySql.Migrations.ModelSnapshotMySqlTest+DataTypesVariable", b =>
     {
         b.Property<int>("Id")
             .ValueGeneratedOnAdd();

         b.Property<byte[]>("TypeByteArray")
             .IsRequired();

         b.Property<byte[]>("TypeByteArray255")
             .IsRequired()
             .HasMaxLength(255);

         b.Property<byte[]>("TypeByteArray255N")
             .HasMaxLength(255);

         b.Property<byte[]>("TypeByteArrayN");

         b.Property<string>("TypeJsonArray")
             .IsRequired()
             .HasColumnType("json");

         b.Property<string>("TypeJsonArrayN")
             .HasColumnType("json");

         b.Property<string>("TypeJsonObject")
             .IsRequired()
             .HasColumnType("json");

         b.Property<string>("TypeJsonObjectN")
             .HasColumnType("json");

         b.Property<string>("TypeString")
             .IsRequired();

         b.Property<string>("TypeString255")
             .IsRequired()
             .HasMaxLength(255);

         b.Property<string>("TypeString255N")
             .HasMaxLength(255);

         b.Property<string>("TypeStringN");

         b.HasKey("Id");

         b.ToTable("DataTypesVariable");
     });
```

Comment (FlyingWraptor) [inline test/EFCore.MySql.Tests/Migrations/ModelSnapshotMySqlTest.cs]: It seems like you've started using the CSharpMigrationsGenerator but did not take the tests into consideration, it is now comparing just the builder code to an 'entire file of code'.

Testing:
```csharp
// <auto-generated />
 using System;
 using Microsoft.EntityFrameworkCore;
 using Microsoft.EntityFrameworkCore.Infrastructure;
 using Microsoft.EntityFrameworkCore.Storage.ValueConversion;

 namespace RootNamespace
 {
     [DbContext(typeof(DbContext))]
     partial class Snapshot : ModelSnapshot
     {
         protected override void BuildModel(ModelBuilder modelBuilder)
         {
 #pragma warning disable 612, 618
             modelBuilder
                 .HasAnnotation("ProductVersion", "2.2.0-rtm-35687")
                 .HasAnnotation("Relational:MaxIdentifierLength", 64);

             modelBuilder.Entity("Pomelo.EntityFrameworkCore.MySql.Migrations.ModelSnapshotMySqlTest+DataTypesVariable", b =>
                 {
                     b.Property<int>("Id")
                         .ValueGeneratedOnAdd();

                     b.Property<byte[]>("TypeByteArray")
                         .IsRequired();

                     b.Property<byte[]>("TypeByteArray255")
                         .IsRequired()
                         .HasMaxLength(255);

                     b.Property<byte[]>("TypeByteArray255N")
                         .HasMaxLength(255);

                     b.Property<byte[]>("TypeByteArrayN");

                     b.Property<string>("TypeJsonArray")
                         .IsRequired()
                         .HasColumnType("json");

                     b.Property<string>("TypeJsonArrayN")
                         .HasColumnType("json");

                     b.Property<string>("TypeJsonObject")
                         .IsRequired()
                         .HasColumnType("json");

                     b.Property<string>("TypeJsonObjectN")
                         .HasColumnType("json");

                     b.Property<string>("TypeString")
                         .IsRequired();

                     b.Property<string>("TypeString255")
                         .IsRequired()
                         .HasMaxLength(255);

                     b.Property<string>("TypeString255N")
                         .HasMaxLength(255);

                     b.Property<string>("TypeStringN");

                     b.HasKey("Id");

                     b.ToTable("DataTypesVariable");
                 });
 #pragma warning restore 612, 618
         }
     }
 }
```

vs the old
```csharp
builder
     .HasAnnotation("Relational:MaxIdentifierLength", 64);

 builder.Entity("Pomelo.EntityFrameworkCore.MySql.Migrations.ModelSnapshotMySqlTest+DataTypesVariable", b =>
     {
         b.Property<int>("Id")
             .ValueGeneratedOnAdd();

         b.Property<byte[]>("TypeByteArray")
             .IsRequired();

         b.Property<byte[]>("TypeByteArray255")
             .IsRequired()
             .HasMaxLength(255);

         b.Property<byte[]>("TypeByteArray255N")
             .HasMaxLength(255);

         b.Property<byte[]>("TypeByteArrayN");

         b.Property<string>("TypeJsonArray")
             .IsRequired()
             .HasColumnType("json");

         b.Property<string>("TypeJsonArrayN")
             .HasColumnType("json");

         b.Property<string>("TypeJsonObject")
             .IsRequired()
             .HasColumnType("json");

         b.Property<string>("TypeJsonObjectN")
             .HasColumnType("json");

         b.Property<string>("TypeString")
             .IsRequired();

         b.Property<string>("TypeString255")
             .IsRequired()
             .HasMaxLength(255);

         b.Property<string>("TypeString255N")
             .HasMaxLength(255);

         b.Property<string>("TypeStringN");

         b.HasKey("Id");

         b.ToTable("DataTypesVariable");
     });
```

**Comment by FlyingWraptor** [comment]
All issues regarding the Integration Tests are related to the Pomelo.JsonObject, comparing internally in EF throws the nullreferenceexception for it.

**Comment by FlyingWraptor** [comment]
@caleblloyd  I'm to unfamiliar with the Pomelo.JsonObject and internal comparing (which is the cause for most errors), mind sharing your knowledge towards these errors?

I'd like some feedback so I can continue helping fix the merge.

**Comment by FlyingWraptor** [comment]
> I can try to look into it over the weekend. I pretty much just pick a failed test, run that test exclusively, and use the stack trace and breakpoints to debug.

Thought you might've had more details already. I'll check it out some more

**Comment by caleblloyd** [comment]
The JSON Object NPE was in the Pomelo.JsonObject class, removing the static operator overloads here fixed it:
https://github.com/PomeloFoundation/Pomelo.JsonObject/pull/9/files#diff-55cf6e8e2e1ddc9c92731e5557fafffeL208

The full test suite still has some failing tests after that fix was made, but it seems to me that it may be new incompatibilities between 2.2 running against the 2.1 test suite

```
Total tests: 7426. Passed: 7211. Failed: 23. Skipped: 192.
```

Since TravisCI runs a subset of the tests anyways, I'll try to just get those passing and publish a `2.2.0-preview1`

**Comment by FlyingWraptor** [comment]
> The JSON Object NPE was in the Pomelo.JsonObject class, removing the static operator overloads here fixed it:
> https://github.com/PomeloFoundation/Pomelo.JsonObject/pull/9/files#diff-55cf6e8e2e1ddc9c92731e5557fafffeL208
> 
> The full test suite still has some failing tests after that fix was made, but it seems to me that it may be new incompatibilities between 2.2 running against the 2.1 test suite
> 
> ```
> Total tests: 7426. Passed: 7211. Failed: 23. Skipped: 192.
> ```
> Since TravisCI runs a subset of the tests anyways, I'll try to just get those passing and publish a `2.2.0-preview1`

Yes, some of the errors indeed are, I've placed a review comment related to the CSharpMigrationsGenerator test but have yet to get a response with your opinion about handling the situation.

File: test/EFCore.MySql.Tests/Migrations/ModelSnapshotMySqlTest.cs (hunk 1 above)
Comment (caleblloyd) [reply test/EFCore.MySql.Tests/Migrations/ModelSnapshotMySqlTest.cs]: Yes this change had to be made in 2.2.0, I don't have a sample of how to update the test though.

**Comment by caleblloyd** [comment]
`CSharpMigrationsGenerator` was the only real change I had to make due to a backwards compatible break in 2.2.0.  The 2.2.0 tests have changed significantly though so I don't see an example that can be easily backported to the 2.1.x test suite right now...

I think we'll have to be OK with the low number of test failures and assume that most of them are 2.2.0 vs. 2.1.0 test suite incompatibilities.  If there's any real bugs that should come out when users test `2.2.0-preview1` code.

**Pull request merged**